text/0088-fix-memory-limitiations-in-session-replays-access-pattern.md

```diff
@@ -0,0 +1,414 @@
+- The probability of data loss from an outage goes up.
+  - If there are problems upstream we have to take action to pause the job.
+
+**Questions**
+
+- Bruno Garcia asks: How do we integrate this feature for self-hosted users?
+
+### 7. Upgrade ClickHouse Version to Utilize Experimental Features
+
+**Proposal**
+
+Upgrading to a newer version of ClickHouse will enable us to use experimental features such as "Live View" and "Window View".
+
+**Drawbacks**
+
+- Would require Sentry to manage the ClickHouse installation.
+  - Feature is not present in a stable release.
+  - Feature is not present on any version of ClickHouse in Altinity.
+
+**Questions**
+
+### 8. Use an Alternative OLAP Database
+
+**Proposal**
+
+OLAP Databases such as Apache Pinot support upserts which appear to be a key requirement for the Replays product. An aggregation state schema can be defined for merging columns in real-time https://docs.pinot.apache.org/basics/data-import/upsert.
+
+**Drawbacks**
+
+- Changes to the data model would be necessary. We will not be able to aggregate array columns.
+- Pinot has ordering constraints.
+  - If segments arrive after their successor has already been ingested the aggregated state will not contain those rows.
+- There are no existing Apache Pinot installations within the Sentry org.
+- The Replays team is not large enough or experienced enough to manage a Pinot installation.
+  - This would require another team assuming the burden for us.
+  - Otherwise, additional budget would need to be allocated to the Replays team to hire outside experts.
+- We need to re-write our application logic for querying the datastore.
+- Migration pains.
+
+**Questions**
+
+### 9. Use an Alternative OLTP Database
+
+**Proposal**
+
+OLTP databases such as PostgreSQL and AlloyDB support updates which appear to be a key requirement for the Replays product. Our scale is small enough that a sharded PostgreSQL database could handle it. Read volume is low relative to write volume. We could optimize our database for this use case.
+
+**Drawbacks**
+
+- Race conditions will require single-threaded processing of like replay-id events.
+- Duplicate messages will necessitate ordering requirements.
+- Always a possibility for dropped and duplicated data regardless of safe guards.
+- AlloyDB is still in developer preview on Google Cloud.
+- We need to re-write our application logic for querying the datastore.
+- Migration pains.
+
+**Questions**
+
+- Colton Allen asks: Is it possible to write to tables partitioned by the hour? When the hour expires the table is indexed for fast read performance. Replays on the most recent hour partition would be unavailable while we are writing to it. Does PostgreSQL expose behavior that allows us to query over multiple partitions transparently? Is this even necessary given our scale? Currently processing 200 messages per second.
+
+### 10. Manually Manage An Aggregated Materialized View
```

Comment:
Infra burden: medium (cron job generated inserts is a new mechanism)
Maintainability: Medium
Customer Impact: Low/medium (depending on if the product changes)

Questions:
Do we need to merge the last hour segments with aggregated?